swig/csharp/CMakeLists.txt

```diff
@@ -9,275 +15,753 @@ set(GDAL_CSHARP_INSTALL_DIR
-  add_test(NAME csharp_GDALOverviews
-           COMMAND ${CSHARP_INTERPRETER} GDALOverviews.exe
-                   ${CMAKE_CURRENT_SOURCE_DIR}${_separator}Data${_separator}sample.tif NEAREST 2 4)
-  add_test(NAME csharp_GDALInfo COMMAND ${CSHARP_INTERPRETER} GDALInfo.exe
-                                        ${CMAKE_CURRENT_SOURCE_DIR}${_separator}Data${_separator}sample.tif)
-  add_test(NAME csharp_OSRTransform COMMAND ${CSHARP_INTERPRETER} OSRTransform.exe)
-
-  foreach (
-    test_name IN
-    ITEMS csharp_createdata
-          csharp_ogrinfo
-          csharp_GDALWrite
-          csharp_GDALDatasetWrite
-          csharp_GDALCreateCopy
-          csharp_GDALOverviews
-          csharp_GDALInfo
-          csharp_OSRTransform)
-    set_property(TEST ${test_name} PROPERTY ENVIRONMENT "${TEST_ENV}")
-  endforeach ()
+gdal_build_csharp_sample(
+  OUTPUT
+  OGRLayerAlg.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/OGRLayerAlg.cs
+  DEPENDS
+  osr_csharp
+  ogr_csharp
+  SYSTEM_DEPENDS
+  OSGeo.OGR
+  OSGeo.OSR)
 
-endif ()
+gdal_build_csharp_sample(
+  OUTPUT
+  OGRFeatureEdit.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/OGRFeatureEdit.cs
+  DEPENDS
+  osr_csharp
+  ogr_csharp
+  SYSTEM_DEPENDS
+  OSGeo.OGR
+  OSGeo.OSR)
+
+gdal_build_csharp_sample(
+  OUTPUT
+  GDALWarp.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/GDALWarp.cs
+  DEPENDS
+  gdal_csharp
+  SYSTEM_DEPENDS
+  OSGeo.GDAL)
+
+gdal_build_csharp_sample(
+  OUTPUT
+  GetCRSInfo.exe
+  SOURCE
+  ${CMAKE_CURRENT_SOURCE_DIR}/apps/GetCRSInfo.cs
+  DEPENDS
+  osr_csharp
+  SYSTEM_DEPENDS
+  OSGeo.OGR)
+
+# Custom Target to make the C# wrap Libraries and use SWIG to build the .cs files
+add_custom_target(csharp_interface DEPENDS gdalconst_wrap osr_wrap ogr_wrap gdal_wrap)
+
+# Custom Target to make the c# bindings - will be run as part of ALL but can also be built seperately
+add_custom_target(csharp_binding ALL DEPENDS csharp_interface OSGeo.GDAL.CONST OSGeo.GDAL OSGeo.OGR OSGeo.OSR)
+
+# Custom Target to build the C# bindings and the sample binaries
+add_custom_target(
+  csharp_samples
+  DEPENDS csharp_binding
+          OgrInfo.exe
+          CreateData.exe
+          OSRTransform.exe
+          GDALRead.exe
+          GDALReadDirect.exe
+          GDALAdjustContrast.exe
+          GDALDatasetRasterIO.exe
+          GDALWrite.exe
+          GDALDatasetWrite.exe
+          GDALColorTable.exe
+          WKT2WKB.exe
+          OGRGEOS.exe
+          ReadXML.exe
+          GDALInfo.exe
+          GDALOverviews.exe
+          GDALCreateCopy.exe
+          GDALGetHistogram.exe
+          GDALTest.exe
+          OGRLayerAlg.exe
+          GDALWarp.exe
+          OGRFeatureEdit.exe
+          GetCRSInfo.exe)
+
+# set up the tests
+
+include(GdalSetTestEnv)
+gdal_set_test_env(TEST_ENV)
+
+add_test(
+  NAME csharp_createdata
+  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+  COMMAND ${CSHARP_INTERPRETER} CreateData/CreateData Data pointlayer)
```

Comment:
I'm wondering how the Data subdirectory is found, as it is normally in the source tree and here the working directory is the binary directory. Is there some copying somewhere ? I can't see under the swig/csharp directory of my build tree.

Another issue I have on Ubuntu 20:04 with mcs --version == Mono C# compiler version 6.8.0.105 is that tests fail to run (the version on GitHub actions seems to be a non-stock one: "Found the following C# Mono versions: 6.12.0")
The errors I get are like:
```
test 16
      Start 16: csharp_createdata

16: Test command: /usr/bin/mono "CreateData/CreateData" "Data" "pointlayer"
16: Environment variables: 
16:  LD_LIBRARY_PATH=/home/even/gdal/gdal/build_cmake:/home/even/idb_client/lib/esql:/home/even/FileGDB_API-RHEL7-64gcc83/lib:/home/even/install-mongocxx-3.6.5/lib:/home/even/gdal/gdal/build_cmake:/home/even/spatialite/install-libspatialite-5.0.1/lib:/home/even/MrSID_DSDK-9.5.4.4709-rhel6.x86-64.gcc531/Raster_DSDK/lib:/home/even/MrSID_DSDK-9.5.4.4709-rhel6.x86-64.gcc531/Lidar_DSDK/lib:/home/even/kakadu/v8_2-00679N:/home/even/instantclient_12_1:
16:  GDAL_DRIVER_PATH=/home/even/gdal/gdal/build_cmake/gdalplugins
16:  GDAL_DATA=/home/even/gdal/gdal/data
16: Test timeout computed to be: 1500
16: Cannot open assembly 'CreateData/CreateData': No such file or directory.
 1/13 Test #16: csharp_createdata ................***Failed    0.01 sec
```
And indeed there's no CreateData/CreateData file, but an empty CreateData directory:
```
even@even-ThinkPad-P15v-Gen-1:~/gdal/gdal/build_cmake$ ll swig/csharp/CreateData.exe 
-rwxrwxr-x 1 even even 7168 janv. 19 11:29 swig/csharp/CreateData.exe*
even@even-ThinkPad-P15v-Gen-1:~/gdal/gdal/build_cmake$ ll swig/csharp/CreateData
total 8
drwxrwxr-x  2 even even 4096 janv. 19 11:28 ./
drwxrwxr-x 29 even even 4096 janv. 19 11:29 ../
```

Reply:
The Data directory for these tests is (and was in the legacy scripts) created by the CreateData executable in the working directory.

That is also - for information - why I needed to add dependencies to the tests - they need to depend on CreateData completing and given the multiple process nature of cmake and ctest there were problems with some tests starting before Data was created.

Reply:
The problem with Mono -- I think I know what that is. The directory structure is slightly different when building with Mono. I need to add something there.

Reply:
Mono Tests fixed and working and ctest running on the Mono test build